File: php/integration_tests.php
<?php
/**
 * Weiyun PHP Integration Tests
 * Mirrors python/integration_tests.py
 *
 * Usage:
 *   WEIYUN_MCP_TOKEN=xxx php integration_tests.php [--test list|upload|download|delete|all]
 *
 * Requires a folder named "CI" at the root of your Weiyun drive.
 */

require_once __DIR__ . '/src/Weiyun/Client.php';
use Weiyun\Client;

// ─── Bootstrap ───────────────────────────────────────────────────────────────

$token = getenv('WEIYUN_MCP_TOKEN');
if (!$token) {
    fwrite(STDERR, "Error: WEIYUN_MCP_TOKEN is not set.\n");
    exit(1);
}

$mcpUrl = getenv('WEIYUN_MCP_URL') ?: 'https://www.weiyun.com/api/v3/mcpserver';
$client = new Client($token, $mcpUrl);

// ─── Helpers ─────────────────────────────────────────────────────────────────

function getRootPdir(Client $client): string {
    $res = $client->listFiles(1);
    return $res['pdir_key'] ?? '';
}

function getCiDirKey(Client $client): string {
    $offset = 0;
    $limit  = 50;
    while (true) {
        $res     = $client->listFiles($limit, 0, $offset);
        $dirList = $res['dir_list'] ?? [];
        foreach ($dirList as $d) {
            if (($d['dir_name'] ?? '') === 'CI') {
                echo "✅ Found /CI directory: {$d['dir_key']}\n";
                return $d['dir_key'];
            }
        }
        if ($res['finish_flag'] ?? true) {
            break;
        }
        $offset += $limit;
    }
    throw new RuntimeException(
        "⚠️  /CI directory not found. Please create a folder named 'CI' " .
        "in the root of your Weiyun drive for integration tests to work."
    );
}

// ─── Tests ───────────────────────────────────────────────────────────────────

$passed = 0;
$failed = 0;

function run_test(string $name, callable $fn): void {
    global $passed, $failed;
    echo "\n▶ $name\n";
    try {
        $fn();
        $passed++;
    } catch (Throwable $e) {
        echo "❌ FAILED: " . $e->getMessage() . "\n";
        $failed++;
    }
}

function assert_has_key(array $arr, string $key): void {
    if (!array_key_exists($key, $arr)) {
        throw new RuntimeException("Expected key '{$key}' not found in response.");
    }
}

// Locate the /CI directory once
$ciDirKey  = getCiDirKey($client);
$rootPdir  = getRootPdir($client);

// ── test_list_ci_dir ─────────────────────────────────────────────────────────
function test_list_ci_dir(Client $client, string $ciDirKey, string $rootPdir): void {
    $res = $client->listFiles(10, 0, 0, $ciDirKey, $rootPdir);
    assert_has_key($res, 'file_list');
    assert_has_key($res, 'dir_list');
    echo "✅ Successfully listed files in /CI directory.\n";
}

// ── test_upload_file ─────────────────────────────────────────────────────────
function test_upload_file(Client $client, string $ciDirKey): void {
    $filename = "test_php_upload_" . time();
    file_put_contents($filename, "Hello from CI Upload Test!");
    try {
        $res = $client->upload($filename, $ciDirKey);
        assert_has_key($res, 'file_id');
        echo "✅ Successfully uploaded {$filename} to /CI.\n";
    } finally {
        if (file_exists($filename)) unlink($filename);
    }
}

// ── test_download_file ───────────────────────────────────────────────────────
function test_download_file(Client $client, string $ciDirKey): void {
    $filename = "tencent-weiyun.zip";
    file_put_contents($filename, "Dummy ZIP content for download testing");
    $upRes  = $client->upload($filename, $ciDirKey);
    $fileId = $upRes['file_id'];

    $dlRes = $client->download([["file_id" => $fileId, "pdir_key" => $ciDirKey]]);
    assert_has_key($dlRes, 'items');
    if (count($dlRes['items']) === 0) {
        throw new RuntimeException("Download items list is empty.");
    }
    assert_has_key($dlRes['items'][0], 'https_download_url');
    echo "✅ Successfully obtained real download link for {$filename}.\n";

    if (file_exists($filename)) unlink($filename);
}

// ── test_delete_file ─────────────────────────────────────────────────────────
function test_delete_file(Client $client, string $ciDirKey): void {
    $filename = "test_php_delete_" . time() . ".txt";
    file_put_contents($filename, "Will automatically delete me!");
    try {
        $upRes  = $client->upload($filename, $ciDirKey);
        $fileId = $upRes['file_id'];

        $delRes = $client->delete(
            [["file_id" => $fileId, "pdir_key" => $ciDirKey]],
            null,
            true
        );
        assert_has_key($delRes, 'freed_index_cnt');
        if ($delRes['freed_index_cnt'] < 1) {
            throw new RuntimeException("freed_index_cnt should be >= 1.");
        }
        echo "✅ Successfully deleted {$filename} from /CI.\n";
    } finally {
        if (file_exists($filename)) unlink($filename);
    }
}

// ─── Argument Parsing & Runner ───────────────────────────────────────────────

$opts  = getopt('', ['test:']);
$which = $opts['test'] ?? 'all';

if (in_array($which, ['list',     'all'])) run_test('test_list_ci_dir',
    fn() => test_list_ci_dir($client, $ciDirKey, $rootPdir));
if (in_array($which, ['upload',   'all'])) run_test('test_upload_file',
    fn() => test_upload_file($client, $ciDirKey));
if (in_array($which, ['download', 'all'])) run_test('test_download_file',
    fn() => test_download_file($client, $ciDirKey));
if (in_array($which, ['delete',   'all'])) run_test('test_delete_file',
    fn() => test_delete_file($client, $ciDirKey));

echo "\n────────────────────────────────\n";
echo "Results: {$passed} passed, {$failed} failed\n";
exit($failed > 0 ? 1 : 0);

File: python/integration_tests.py
import unittest
import argparse
import sys
import os
import time

from weiyun_api import WeiyunClient

# Real integration tests hitting the Weiyun API endpoint.
# Requires WEIYUN_MCP_TOKEN environment variable.

class TestRealIntegration(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        token = os.environ.get("WEIYUN_MCP_TOKEN")
        if not token:
            print("Error: WEIYUN_MCP_TOKEN is not set.")
            sys.exit(1)
        cls.client = WeiyunClient(token=token)
        cls.ci_dir_key = cls._get_ci_dir()

    @classmethod
    def _get_root_pdir(cls):
        res = cls.client.list_files(limit=1)
        return res.get("pdir_key", "")

    @classmethod
    def _get_ci_dir(cls):
        """Finds the /CI directory in the root. Raises an error if not found."""
        offset = 0
        limit = 50
        while True:
            res = cls.client.list_files(limit=limit, offset=offset)
            dir_list = res.get("dir_list", [])
            for d in dir_list:
                if d.get("dir_name") == "CI":
                    print(f"✅ Found /CI directory: {d['dir_key']}")
                    return d["dir_key"]
            if res.get("finish_flag", True):
                break
            offset += limit
            
        raise RuntimeError("⚠️ /CI directory not found. Please create a folder named 'CI' in the root of your Weiyun drive for integration tests to work.")

    def test_list_ci_dir(self):
        res = self.client.list_files(limit=10, dir_key=self.ci_dir_key, pdir_key=self.__class__._get_root_pdir())
        self.assertIn("file_list", res)
        self.assertIn("dir_list", res)
        print("✅ Successfully listed files in /CI directory.")

    def test_upload_file(self):
        filename = f"test_python_upload_{int(time.time())}"
        with open(filename, "w") as f:
            f.write("Hello from CI Upload Test!")
        
        try:
            res = self.client.upload(file_path=filename, pdir_key=self.ci_dir_key)
            self.assertIn("file_id", res)
            print(f"✅ Successfully uploaded {filename} to /CI.")
        finally:
            if os.path.exists(filename):
                os.remove(filename)

    def test_download_file(self):
        # We must ensure tencent-weiyun.zip exists in /CI, or upload it first
        filename = "tencent-weiyun.zip"
        with open(filename, "w") as f:
            f.write("Dummy ZIP content for download testing")
            
        up_res = self.client.upload(file_path=filename, pdir_key=self.ci_dir_key)
        file_id = up_res["file_id"]
        
        # Download
        dl_res = self.client.download([{"file_id": file_id, "pdir_key": self.ci_dir_key}])
        self.assertIn("items", dl_res)
        self.assertGreater(len(dl_res["items"]), 0)
        self.assertIn("https_download_url", dl_res["items"][0])
        print(f"✅ Successfully obtained real download link for {filename}.")
        
        if os.path.exists(filename):
            os.remove(filename)

    def test_delete_file(self):
        filename = f"test_python_delete_{int(time.time())}.txt"
        with open(filename, "w") as f:
            f.write("Will automatically delete me!")
            
        try:
            up_res = self.client.upload(file_path=filename, pdir_key=self.ci_dir_key)
            file_id = up_res["file_id"]
            
            # Delete file using weiyun.delete
            del_res = self.client.delete(file_list=[{"file_id": file_id, "pdir_key": self.ci_dir_key}], delete_completely=True)
            self.assertIn("freed_index_cnt", del_res)
            self.assertGreaterEqual(del_res["freed_index_cnt"], 1)
            print(f"✅ Successfully deleted {filename} from /CI.")
        finally:
            if os.path.exists(filename):
                os.remove(filename)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Run specific integration test")
    parser.add_argument("--test", choices=["list", "upload", "download", "delete", "all"], default="all")
    args = parser.parse_args()

    suite = unittest.TestSuite()
    if args.test == "list" or args.test == "all":
        suite.addTest(TestRealIntegration('test_list_ci_dir'))
    if args.test == "upload" or args.test == "all":
        suite.addTest(TestRealIntegration('test_upload_file'))
    if args.test == "download" or args.test == "all":
        suite.addTest(TestRealIntegration('test_download_file'))
    if args.test == "delete" or args.test == "all":
        suite.addTest(TestRealIntegration('test_delete_file'))

    runner = unittest.TextTestRunner(verbosity=2)
    result = runner.run(suite)
    sys.exit(not result.wasSuccessful())
